File: Ahorn/entities/DreamFallingBlock.jl
module CommunalHelperDreamFallingBlock
using ..Ahorn, Maple

@mapdef Entity "CommunalHelper/DreamFallingBlock" DreamFallingBlock(x::Integer, y::Integer, 
	width::Integer=Maple.defaultBlockWidth, height::Integer=Maple.defaultBlockHeight,
	featherMode::Bool = false, oneUse::Bool = false)
        
            
const placements = Ahorn.PlacementDict(
    "Dream Falling Block (Communal Helper)" => Ahorn.EntityPlacement(
        DreamFallingBlock,
		"rectangle"
    )
)

Ahorn.minimumSize(entity::DreamFallingBlock) = 8, 8
Ahorn.resizable(entity::DreamFallingBlock) = true, true

function Ahorn.selection(entity::DreamFallingBlock)
    x, y = Ahorn.position(entity)

    width = Int(get(entity.data, "width", 8))
    height = Int(get(entity.data, "height", 8))
	
    return Ahorn.Rectangle(x, y, width, height)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::DreamFallingBlock, room::Maple.Room)
    x = Int(get(entity.data, "x", 0))
    y = Int(get(entity.data, "y", 0))

    width = Int(get(entity.data, "width", 32))
    height = Int(get(entity.data, "height", 32))

    Ahorn.CommunalHelper.renderCustomDreamBlock(ctx, 0, 0, width, height, get(entity.data, "featherMode", false), get(entity.data, "oneUse", false))
end

end

File: Loenn/entities/dream_blocks/dream_falling_block.lua
local drawableRectangle = require("structs.drawable_rectangle")
local communalHelper = require("mods").requireFromPlugin("libraries.communal_helper")

local dreamFallingBlock = {}

dreamFallingBlock.name = "CommunalHelper/DreamFallingBlock"
dreamFallingBlock.fieldInformation = {
    refillCount = {
        fieldType = "integer"
    },
    fallDistance = {
        minimumValue = 0,
        fieldType = "integer"
    }
}

function dreamFallingBlock.depth(room, entity)
    return entity.below and 5000 or -11000
end

-- let's not show the 'chained' field (which is needed in the c# side of the entity)
-- you get to chose whether this falling block is chained the the placements
-- also, because of this, we need to explicitly ignore other options hidden by default
dreamFallingBlock.ignoredFields = {
    "chained",
    "_id",
    "_name"
}

dreamFallingBlock.placements = {
    {
        name = "normal",
        placementType = "rectangle",
        data = {
            width = 16,
            height = 16,
            featherMode = false,
            oneUse = false,
            refillCount = -1,
            below = false,
            quickDestroy = false,
            noCollide = false,
            forceShake = false,
            chained = false
        }
    },
    {
        name = "chained",
        placementType = "rectangle",
        data = {
            width = 16,
            height = 16,
            featherMode = false,
            oneUse = false,
            refillCount = -1,
            below = false,
            quickDestroy = false,
            noCollide = false,
            forceShake = false,
            fallDistance = 64,
            centeredChain = false,
            chainOutline = true,
            indicator = false,
            indicatorAtStart = false,
            chained = true,
            chainTexture = "objects/CommunalHelper/chains/chain"
        }
    }
}

function dreamFallingBlock.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 8, entity.height or 8

    local sprites = {}

    if entity.chained then
        local fallDistance = entity.fallDistance or 16
        local rect = drawableRectangle.fromRectangle("line", x, y, width, height + fallDistance, {1, 1, 1, 0.5})
        rect.depth = 0
        table.insert(sprites, rect)
    end

    table.insert(sprites, communalHelper.getCustomDreamBlockSprites(x, y, width, height, entity.featherMode, entity.oneUse))

    return sprites
end

return dreamFallingBlock
